executor/instruction/iid_executor.rs

```diff
@@ -0,0 +1,138 @@
+    match res {
+        Ok(value) => Ok(Tuple::Single([value])),
+        Err(err) => Err(err),
+    }
+}
+
+impl IidExecutor {
+    pub(crate) fn new(
+        iid: IidInstruction<ExecutorVariable>,
+        variable_modes: VariableModes,
+        _sort_by: ExecutorVariable,
+    ) -> Self {
+        let IidInstruction { iid, types, checks } = iid;
+
+        let var = iid.var().as_variable().unwrap();
+        let output_tuple_positions = TuplePositions::Single([Some(var)]);
+        let checker = Checker::<VariableValue<'_>>::new(checks, HashMap::from_iter([(var, EXTRACT_LHS)]));
+
+        let filter_fn: Arc<IidFilterFn> = Arc::new(move |res| match res {
+            Ok(value) => Ok(types.contains(&value.as_thing().type_())),
+            Err(err) => Err(err.clone()),
+        });
+
+        Self { iid, variable_modes, tuple_positions: output_tuple_positions, filter_fn, checker }
+    }
+
+    pub(crate) fn get_iterator(
+        &self,
+        context: &ExecutionContext<impl ReadableSnapshot + 'static>,
+        row: MaybeOwnedRow<'_>,
+    ) -> Result<TupleIterator, Box<ConceptReadError>> {
+        let filter = self.filter_fn.clone();
+        let check = self.checker.filter_for_row(context, &row);
+        let filter_for_row: Box<IidFilterFn> = Box::new(move |item| match filter(item) {
+            Ok(true) => check(item),
+            fail => fail,
+        });
+
+        let snapshot = &**context.snapshot();
+        let thing_manager = context.thing_manager();
+
+        let iid_parameter = self.iid.iid().as_parameter().unwrap();
+        let bytes = context.parameters().iid(iid_parameter).unwrap();
+
+        let instance = if let Some(object) = ObjectVertex::try_from_bytes(bytes) {
+            let object = Object::new(object);
+            thing_manager
+                .instance_exists(snapshot, object.as_reference())
+                .map(move |exists| exists.then_some(VariableValue::Thing(object.into_owned().into())))
+        } else if let Some(attribute) = AttributeVertex::try_from_bytes(bytes) {
+            let attribute = Attribute::new(attribute);
+            thing_manager
+                .instance_exists(snapshot, attribute.as_reference())
+                .map(move |exists| exists.then_some(VariableValue::Thing(attribute.into_owned().into())))
+        } else {
+            Ok(None)
+        };
+
+        fn as_tuples<T>(it: T) -> Map<T, IidToTupleFn, AsHkt![TupleResult<'_>]>
+        where
+            T: for<'a> LendingIterator<Item<'a> = Result<VariableValue<'a>, Box<ConceptReadError>>>,
+        {
+            it.map::<TupleResult<'_>, IidToTupleFn>(iid_to_tuple)
+        }
+
+        let iterator = AsNarrowingIterator::new(instance.transpose());
+        let as_tuples = as_tuples(iterator.try_filter::<_, IidFilterFn, VariableValue<'_>, _>(filter_for_row));
+        Ok(TupleIterator::Iid(SortedTupleIterator::new(as_tuples, self.tuple_positions.clone(), &self.variable_modes)))
```

Comment:
Flow:
- decode IID to instance,
- verify instance exists in thing manager,
- verify instance is of one of allowed types,
- run other checks.

We need to check the type for occasions such as `match $x iid 0x...; $x has $a;` where the iid is an attribute id, so that we can discard early. Otherwise `has` would crash since it expects an object at `$x`.

Reply:
Are the checks done at the end or at the beginning?